src/core/coap/secure_coap_server.cpp

```diff
@@ -117,6 +117,8 @@ void SecureServer::Receive(Message &aMessage, const Ip6::MessageInfo &aMessageIn
     {
         mPeerAddress.SetPeerAddr(aMessageInfo.GetPeerAddr());
         mPeerAddress.SetPeerPort(aMessageInfo.GetPeerPort());
+        mPeerAddress.SetSockAddr(aMessageInfo.GetSockAddr());
```

Comment:
Good catch.  These will generally be unicast addresses.  But to be safe/complete, we should check whether this is a unicast address before setting this value.  If the address is multicast, we should not perform this logic and rely on default source address selection.